### ISSUE TITLE
failed mapping AST context canceled

### ISSUE THREAD

**Issue by RobbanHoglund** [grafana/loki#14384]
**Describe the bug**
After upgrade to Loki 3.2.0 it is repeatedly logging this for some queries:
ts=2024-10-04T01:09:28.649297614Z caller=spanlogger.go:111 middleware=QueryShard.astMapperware org_id=fake traceID=35d59f6204919421 user=fake caller=log.go:168 level=warn msg="failed mapping AST" err="context canceled" query="{application=\"myapp\",level=~\"INFO|WARN|ERROR\"} "

**To Reproduce**
1. Started Loki 3.2.0
2. From Grafana explore do {application="myapp",level=~"INFO|WARN|ERROR|TRACE"}

**Expected behavior**
If this is an actual problem in Loki 3.2.0  it would be good to get more information about what is the root cause.

**Environment:**
 - Infrastructure: bare-metal
 - Deployment tool: bash scripts

**Screenshots, Promtail config, or terminal output**
If applicable, add any output to help explain your problem.
![Image](https://github.com/user-attachments/assets/3d2cbc0f-9675-4745-a9c0-cfe777976710)
![Image](https://github.com/user-attachments/assets/f40652c6-94c3-4fcd-8503-6eb290f65e11)

**Comment by jammiemil** [comment]
Also seeing this in a cluster running 3.2.0, however its only happening on one of our Read pods (Running SSD via Helm), the same pod is reporting High query latency (upto 50s), and reporting a fair amount of context cancelled in the logs, All the other read pods are behaving just fine and returning query in sub-seconds

**Comment by eplightning** [comment]
I ran into something similar with 3.1.1 . Restarting the broken pod seem to have fixed that.

**Comment by vincentnonim** [comment]
I'm having the same issue with Loki 3.2.0 deployed using Docker and SSD.
Once in a while Read container return msg="failed mapping AST" err="context canceled" errors, only way to resolve is to restart the read container. I don't spot any issues in metrics...

**Comment by someStrangerFromTheAbyss** [comment]
+1 here. When it happens, seems more severe on one specifically more then the others.

**Comment by JeffreyVdb** [comment]
Restarting the pod works for me as well.

**Comment by Khagesh16** [comment]
Same issue in my case, deployed loki in simple scalable mode. When I run the query, I keep seeing this error.

**Comment by someStrangerFromTheAbyss** [comment]
Ok so i restarted the read pods, still get the error, but no more EOF errors/time outs from query. It seems a good way to detect if the impact is noticable on client's perspective is to check if there is any 499 HTTP code error returned from the nginx. There is probably some loki metrics somewhere that indicate it, but i could not find it.

**Comment by someStrangerFromTheAbyss** [comment]
I did some digging, and my initial problems of EOF/timeout is not linked to that warning. I tried to removed a bunch of thing from my config hoping to remove this warning, could not find any way to remove it. Remove my old store config, did not make it dissapear. Looking at the code, it seems to happen when parsing the query and sharding the query. 

Will sadly stop the investigation here since my query problems does not seem to be linked to this warning.

**Comment by emadolsky** [comment]
+1 we also see this issue occasionally and deleting the affected pod fixes the issue.

**Comment by BlexToGo** [comment]
We run loki `3.1.1` with `deploymentMode: SimpleScalable` and also saw the same error. Restarting the `loki-read` deployment fixed the issue in seconds.

**Comment by marcotuna** [comment]
I'm experiencing the same issue. I have a few clusters running Loki 3.2.1 in Simple Scalable Mode (SSD). All clusters display the warning "failed mapping AST," though it doesn’t seem to impact Loki’s functionality, except in two specific clusters. In these cases, similar to previous reports, one of the Pods stops processing queries and only resumes normal operation after a restart.

This issue began occurring only after upgrading from version 2.9.X to 3.2.X, following the guide for upgrading to version 3.X.

As @someStrangerFromTheAbyss mentioned, there is a noticeable increase in 499 HTTP status codes, which can be seen on the "Loki Operational" Grafana Dashboard.

In my observations, the affected Pod still receives requests from Nginx (loki-gateway) but doesn’t return a response, appearing to get stuck while processing. My deployment relies on AWS S3.

**Comment by someStrangerFromTheAbyss** [comment]
For the http code 499, i can confirm these are not linked. I was able to resolve the 499 issue seperately then that error message. I made a really big answer on the 499 issue here:
https://github.com/grafana/loki/issues/7084

TLDR: The 499 issue is linked to the boltdb-shipper indexer and connection to the backend when using boltdb-shipper. If you can remove these, the 499 should dissapear.

**Comment by EZ4BRUCE** [comment]
any update of this issue ?

**Comment by LeszekBlazewski** [comment]
We are observing this issue on loki 3.3.0 deployed via helm chart version: 6.22.0 as well. We are running loki in SimpleScalable deployment. We see a lot of those logs when using [Explore logs](https://grafana.com/docs/grafana-cloud/visualizations/simplified-exploration/logs/#explore-logs) feature on grafana v11.3.1 (64b556c137):  (we have all of the prerequisites on loki & grafana configured).

**Comment by someStrangerFromTheAbyss** [comment]
After some more digging, its seems this error is linked to the 499 error indeed.

Seems to only happen to one of the read pods after a deployment and happens most of the time with the first read pods that connect to the backend pod. Im back at investiguating the issue since this affects client queries.

**Comment by disparo13** [comment]
I'm experiencing the same problem. I'm running Loki version 3.2.0, and at some moments, I cannot load Label filters in Grafana -> Explore, getting an infinite spinner in the browser's Label filter field and "failed mapping AST" in loki-read pods. Restart helps for a while, but then the same situation reoccurs.

**Comment by emadolsky** [comment]
@someStrangerFromTheAbyss I can verify that it's a connectivity issue between reads and backends, and it happens as well when you just rollout the backend pods.

**Comment by Starefossen** [comment]
We have this issues from time to time as well. No noticable change upgrading to Loki v3.3.0.

**Comment by Khagesh16** [comment]
Bump, stuck on this issue since months. Even restart does not seem to be helping.

**Comment by garam42** [comment]
I am currently using the following components. When I use the Grafana Explore log feature and refresh, this error keeps occurring. Has this issue been resolved?

promtail: 3.0.0
loki: 3.3.2 (single binary)
grafana: 11.3"

**Comment by r-asiebert** [comment]
One more with loki 3.1.1 on "read" pods, lots of:
```
ts=... caller=spanlogger.go:109 middleware=QueryShard.astMapperware org_id=... traceID=622160cf96ed4ead user=... level=warn msg="failed mapping AST" err="context canceled" query="..."
```

This happened after taking the one "backend" pod offline for maintenance for a few minutes. Restarting the "read" deployment worked; no bloom feature enabled here.

**Comment by tuananh** [comment]
we're facing the same issue. any pointers here? restarting read deployment fixes it.

**Comment by patsevanton** [comment]
same issue with helm chart loki version 6.27.0

**Comment by scottsananderson** [comment]
We are also experiencing this issue with via Helm chart 6.28.0 , using SSD.  Restarting the loki-read pods resolves the issue.  Problem occurs at least once a month.

**Comment by towolf** [comment]
Is a livenessProbe able to solve this problem too?

**Comment by someStrangerFromTheAbyss** [comment]
@towolf i was thinking to also change the liveness probe to check /loki/api/v1/labels in the last minute to check if the read pod is connected correctly

**Comment by quyenhoang96** [comment]
+1. Is there any update to this ticket ?

**Comment by RobbanHoglund** [comment]
Did you see this issue @cyriltovena  ?

**Comment by mauriciocampos** [comment]
+1. Also facing the same issues here and a restart on read does fix the issue.

**Comment by kacperpabian** [comment]
+1 version 3.4.2 issue still appears. We have multiple big loki clusters, some of them are experiencing this issue. Error occurs often but only after some time is blocking most of the queries, restarting read deployment helps, but for some time.

**Comment by Ravio1i** [comment]
We've faced this issue and were able to resolve this by increasing `max_concurrent`

```yaml
    querier:
      max_concurrent: 8  # from 3
```

**Comment by kacperpabian** [comment]
@Ravio1i Do you line it up with cpu limit on read instances?

**Comment by Ravio1i** [comment]
> [@Ravio1i](https://github.com/Ravio1i) Do you line it up with cpu limit on read instances?

In this case the resource requests nor limits were not the issue, in the past we had loki consume lots and increased it but now its fine with:

```yaml
        resources:
          requests:
            cpu: "500m"
            memory: "500Mi"
          limits:
            cpu: "1"
            memory: "1500Mi"
```

**Comment by kacperpabian** [comment]
Ok, increasing max_concurrent or tweaking it for cpu limits does not solve the issue for me. Issue still stands.

**Comment by Ravio1i** [comment]
> Ok, increasing max_concurrent or tweaking it for cpu limits does not solve the issue for me. Issue still stands.

In the past we've also increased:

```yaml
limits_config:
  query_timeout: 600s  # from 300s
```

**Comment by emadolsky** [comment]
Just wanted to share here that in our case, after migrating from SSD mode to distributed (microservice) mode, the problem was resolved.

**Comment by jyotir-aditya-giri** [comment]
We do see issue in distributed mode with `querier.max_concurrent: 16`
Sometimes, restarting doesn't resolve it either and it happens to be auto-resolved.

We run two loki connected to the same bucket with one loki scaled down to 0 replicas and when this issue shows up, we scale up the stand-by loki to allow querying for logs until the primary loki auto-resolves.

**Comment by kacperpabian** [comment]
Unfortunately happens to me also. Changed from SimpleScalable to Distributed and error AST mapping still is happening in frontend-querier. We are using redis as our cache, not sure if that is something that might be related to the issue.

**Comment by tilmanschweitzer** [comment]
We have the same issue. It causes DatasourceErrors in Loki based alarms what is quite annoying. As described by others, restarting the pod solved the issue - until it appears again. It seems to be persistent if it appears in one pod and it seems like all queries to the affected pod fail.

**Possible mitigation: livenessProbe**

I checked that even very simple queries fail (in two occasions) via wget right on the pod. They just hung until wget timed out.

```
 wget -qO- "http://localhost:3100/loki/api/v1/query_range?query={service_name=\"loki\"}&since=10s&limit=1" 
```

As next step I try to mitigate the issue with to following livenessProbe: 

```
    livenessProbe:
      failureThreshold: 3
      httpGet:
        path: /loki/api/v1/query_range?query=%7Bservice_name%3D%22loki%22%7D&since=10s&limit=1
        port: http-metrics
        scheme: HTTP
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 5
```

**Further observations**

I wasn't able to find clear correlations with memory or cpu consumption or other observations. It seems like the issue was triggered more often in our cluster when we had less pods running (e. g. scale down from auto scaler), but it still seems a bit random.

In two occasions, the issue started in multiple pods at the same time. This could be a hint that there is some form of event that causes the issue and that it is not completely random.

I use the following query to quantify the issue per minute and pod:

```
sum by(pod) (rate({service_name="loki"} | logfmt | msg="failed mapping AST" | middleware="QueryShard.astMapperware" [$__auto])) * 60
```

<img width="1321" height="336" alt="Image" src="https://github.com/user-attachments/assets/c9b204be-ac5b-4cac-82c4-c3e15a592389" />

***

**Comment by mxab** [comment]
Not sure if it is related, we run Loki in microservice mode and I noticed that there is a race condition when the frontend restarts and reregisters at the query scheduler.  The scheduler doesn't pick up new queries resulting in a timeout on the loki frontend side

I created an issue for that here https://github.com/grafana/loki/issues/19528 and a PR that mitigates it https://github.com/grafana/loki/pull/19866

unfortunately I haven't got any feedback regarding this.

**Comment by sagarnikam123** [comment]
SOLVED:
I am getting same error in Loki `3.6.3` on monolithic (single node) setup. We are using minio locally, inmemory kv store, frontend.encoding: protobuf, chunk_encoding: snappy, tsdb & embedded_cache. 
I change value to from true to `parallelise_shardable_queries: false` and don't show again.

**Comment by kacperpabian** [comment]
> SOLVED: I am getting same error in Loki `3.6.3` on monolithic (single node) setup. We are using minio locally, inmemory kv store, frontend.encoding: protobuf, chunk_encoding: snappy, tsdb & embedded_cache. I change value to from true to `parallelise_shardable_queries: false` and don't show again.

Yes, thank you. I can also confirm it helped my case.

**Comment by raphaelquati** [comment]
I ran into the same warning after upgrading Loki to 3.7.1, but in my case this does not seem to be a regression.

```
2026-04-17 09:01:40.044 | ts=2026-04-17T12:01:40.002722963Z caller=spanlogger.go:152 middleware=QueryShard.astMapperware org_id=fake user=fake level=warn msg="failed mapping AST" err="context canceled" query="count_over_time({namespace=\"prod\", container=\"xxx\"}[6m] \|~ \".*Background encerrado, será processado novamente em.*\")"
```
In my setup, this happened because I occasionally copy default Helm values into my own `values.yaml`.  
With newer Loki versions, some defaults change, and since I didn’t review those changes, `parallelise_shardable_queries` ended up enabled unintentionally.

After removing `parallelise_shardable_queries` from my Helm values, the warning disappeared.

**Comment by niels-hb** [comment]
We're running into the same issue with grafana/loki:3.7.2. Deployment mode SimpleScalable. Helm chart grafana-community/loki v16.0.1.

Manually deleting the Pod also "solves" the issue for a while.

**Comment by stefanjauker** [comment]
+1
